examples/kaleidoscope.rs

```diff
@@ -0,0 +1,1290 @@
+        //     None => 
+        // };
+        let owned_entry = self.fn_value.get_entry_basic_block();
+        let entry = owned_entry.as_ref().or(entry).unwrap();
+        
+        match entry.get_first_instruction() {
+            Some(first_instr) => builder.position_before(&first_instr),
+            None => builder.position_at_end(entry)
+        }
+        
+        builder.build_stack_allocation(&self.context.f64_type(), name)
+    }
+
+    /// Compiles the specified `Expr` into an LLVM `BasicValue`.
+    /// Note: This method currently returns a `Box`, since it returns either `PhiValue` or `FloatValue`.
+    fn compile_expr(&mut self, expr: &Expr) -> Result<FloatValue, &'static str> {
+        match expr {
+            &Expr::Number(nb) => Ok(self.context.f64_type().const_float(nb)),
+
+            &Expr::Variable(ref name) => {
+                match self.variables.get(name.as_str()) {
+                    Some(var) => Ok(self.builder.build_load(&var, name.as_str()).into_float_value()),
+                    None => Err("Could not find a matching variable.")
+                }
+            },
+
+            &Expr::VarIn { ref variables, ref body } => {
+                let mut old_bindings = Vec::new();
+
+                for &(ref var_name, ref initializer) in variables {
+                    let var_name = var_name.as_str();
+
+                    let initial_val = match initializer {
+                        &Some(ref init) => self.compile_expr(&init)?,
+                        &None => self.context.f64_type().const_float(0.)
+                    };
+
+                    let alloca = self.create_entry_block_alloca(var_name, None);
+
+                    self.builder.build_store(&alloca, &initial_val);
+                    
+                    if let Some(old_binding) = self.variables.remove(var_name) {
+                        old_bindings.push(old_binding);
+                    }
+
+                    self.variables.insert(var_name.to_string(), alloca);
+                }
+
+                let body = self.compile_expr(&body)?;
+
+                for binding in old_bindings {
+                    self.variables.insert(binding.get_name().to_str().unwrap().to_string(), binding);
+                }
+
+                Ok(body)
+            },
+
+            &Expr::Binary { op, ref left, ref right } => {
+                if op == '=' {
+                    // handle assignement
+                    let var_name = match left.borrow() {
+                        &Expr::Variable(ref var_name) => var_name,
+                        _ => {
+                            return Err("Expected variable as left-hand operator of assignement.");
+                        }
+                    };
+
+                    let var_val = self.compile_expr(&right)?;
+                    let var = self.variables.get(var_name.as_str()).ok_or("")?;
+
+                    Ok(unsafe { std::mem::transmute(self.builder.build_store(&var, &var_val)) })
```

Comment:
The method returns a `FloatValue`, but `build_store` returns a `PointerValue`. I'm casting the latter to the former here, since the structure is actually a pointer to a float.